### PR TITLE
Fixes: the unit test failures after coroutine version update

### PR DESCRIPTION
**Pull request by AjeshRPai**
## Fixes 
Fixes the failure of the unit tests after the coroutines library update. PR - https://github.com/wordpress-mobile/WordPress-Android/pull/18500

#### On **Kotlin coroutines library version 1.6.4**
If the coroutine inside a function throws an exception, the test passes, and the exception is thrown silently to the console. In some cases, if the test is asserting a return value and there is a side effect function that throws an exception, then the test still passes. 

#### On  **Kotlin coroutines library version 1.7.1**
The exceptions inside the production code will be thrown, regardless of whether the test is passing, making the unit tests fail. 

More details in this thread - https://github.com/Kotlin/kotlinx.coroutines/issues/1205

### To test:
1. Run the unit tests changed in this PR on another branch with coroutines version 1.6.4 
2. Verify tests are failing 
3. Run the unit tests in this PR  
4. Verify tests are passing 

## Regression Notes
1. Potential unintended areas of impact
Incorrect unit test logic 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: WordPress/src/test/java/org/wordpress/android/ui/domains/DomainSuggestionsViewModelTest.kt [hunk 1]
```diff
@@ -136,13 +136,14 @@ class DomainSuggestionsViewModelTest : BaseUnitTest() {
 
     @Test
     fun `site on blogger plan is requesting only dot blog domain suggestions`() = test {
+        whenever(productsStore.fetchProducts(any())).thenReturn(mock())
```
Comment (ovitrif) [inline WordPress/src/test/java/org/wordpress/android/ui/domains/DomainSuggestionsViewModelTest.kt]: I figured out while trying to fix the failures that we need to stub this call for every test, so I changed it into an inline `mock {…}` declaration, then removed all `whenever`'s for it, and then all tests in the class succeeded 🎉 !

I captured my changes into this patch file:

[fixtest.patch](https://github.com/wordpress-mobile/WordPress-Android/files/11706443/fixtest.patch)

To apply it, copy the file into the root directory of your local repo clone, then run in terminal:
```sh
git apply fixtest.patch
```
🙏🏻

**Review by ovitrif — CHANGES_REQUESTED** [review]
Thanks for this work and the reason it started: that you felt brave and proved to be right when tackling the Coroutines update work 🙇🏻 !

I found a quickfix for the issue we were having in `DomainSuggestionsViewModelTest` and suggested a change via a patchfile 👍🏻 

Everything else LGTM, but IIRC there were more failing tests at CI time, probably the PR is not fully ready. For that reason (my uncertainty) I didn't approve it just yet 👍🏻

File: WordPress/src/test/java/org/wordpress/android/ui/domains/DomainSuggestionsViewModelTest.kt (hunk 1 above)
Comment (AjeshRPai) [reply WordPress/src/test/java/org/wordpress/android/ui/domains/DomainSuggestionsViewModelTest.kt]: Hey @ovitrif 

Thanks 🙌🏼 . I have applied the patch and committed it [98e0769](https://github.com/wordpress-mobile/WordPress-Android/pull/18589/commits/98e07694a61baf961d59c63d18edc075215b9960)

Thanks for testing and finding this solution. 

Am not really sure why an inline mock worked to be honest. 🤔

Comment (ovitrif) [reply WordPress/src/test/java/org/wordpress/android/ui/domains/DomainSuggestionsViewModelTest.kt]: @AjeshRPai My speculation is that we didn't fully understand the actual point where the code failed. The error message was so fuzzy that it sort of meant even the compiler didn't know 😂.

The gist I got of it is that it wasn't really the red test that failed, but "an uncaught exception in the previous test". AFAIK Android Studio runs them in random order, so it wasn't really the test written above of the failed one. I went forward and `@Ignore`'d most of them until the ones left no longer failed.

Then randomly (magic intuition I call it, it's a gift I don't always have), I thought: let's rewrite this mock. It ended up fixing the issue 🥳

**Comment by zwarm** [comment]
@AjeshRPai I'm with Ovi on this one. I pulled this branch and tried the first failing test (as reported by CI) `MediaPickerViewModelTest` and I can't get it to fail locally. However the `ActivityLogViewModelTest` do fail locally. 😮‍💨

**Comment by ovitrif** [comment]
> @AjeshRPai I'm with Ovi on this one. I pulled this branch and tried the first failing test (as reported by CI) `MediaPickerViewModelTest` and I can't get it to fail locally. However the `ActivityLogViewModelTest` do fail locally. 😮‍💨

Hey folks 👋🏻 , Fyi I've started on working to fix these tests yesterday, sharing with you in hope we don't duplicate the same efforts 🙏🏻 !

I've managed to find some small fixes that made all tests in `ActivityLogViewModelTest` green, I will continue to the next classes.

Mostly doing this to see if the new coroutines version fixes our mutex crash bug (#18589) 😄

**Review by ovitrif — APPROVED** [review]
Should be ready when CI passes 🤞🏻

**Comment by AjeshRPai** [comment]
Thanks @ovitrif and @zwarm for helping me with this PR  🙌🏼

**Pull request merged**